packages/services/src/activity/activity.service.ts

```diff
@@ -0,0 +1,335 @@
+    ]);
+    const eventsByTxId = getEventsByTxId(txEvents);
+    const sip10AssetTxs = txs.filter(tx =>
+      tx.tx.post_conditions.some(
+        pc =>
+          pc.type === 'fungible' &&
+          getAssetIdentifierFromContract(
+            pc.asset.contract_address,
+            pc.asset.contract_name,
+            pc.asset.asset_name
+          ) === asset.assetId
+      )
+    );
+    const activityList: Activity[] = [];
+    for (const tx of [...pendingTxs, ...sip10AssetTxs.map(tx => tx.tx)]) {
+      const txActivity = await getStacksTxActivity(tx, eventsByTxId.get(tx.tx_id), account);
+      if (txActivity) {
+        activityList.push(
+          ...(await Promise.all(txActivity.map(activity => applyMarketData(activity, signal))))
+        );
+      }
+    }
+    return activityList.sort(sortActivityByTimestampDesc);
+  }
+
+  async function getStacksTxActivity(
+    tx: HiroStacksTransaction | HiroStacksMempoolTransaction,
+    txEvents: HiroTransactionEvent[] = [],
+    account: AccountIdentifier,
+    signal?: AbortSignal
+  ): Promise<OnChainActivity[]> {
+    switch (tx.tx_type) {
+      case 'token_transfer':
+        return mapTokenTransferActivity(tx, account);
+      case 'smart_contract':
+        return mapSmartContractActivity(
+          tx,
+          account,
+          await getStacksAssetTransfersWithInfo(tx, txEvents, signal)
+        );
+      case 'contract_call':
+        return mapContractCallActivity(
+          tx,
+          account,
+          await getStacksAssetTransfersWithInfo(tx, txEvents, signal)
+        );
+      default:
+        return [];
+    }
+  }
+
+  async function getStacksAssetTransfersWithInfo(
+    tx: HiroStacksTransaction | HiroStacksMempoolTransaction,
+    txEvents: HiroTransactionEvent[],
+    signal?: AbortSignal
+  ): Promise<StacksAssetTransferWithInfo[]> {
+    const transfers = getStacksAssetTransfers(tx, txEvents);
+    return (await Promise.all(transfers.map(t => addStacksTransferAssetInfo(t, signal)))).filter(
+      isDefined
+    );
+  }
+
+  async function addStacksTransferAssetInfo(
+    transfer: StacksAssetTransfer,
+    signal?: AbortSignal
+  ): Promise<StacksAssetTransferWithInfo | undefined> {
+    if (transfer.assetId === 'STX') {
+      return {
+        ...transfer,
+        assetInfo: stxCryptoAsset,
+      };
+    } else {
+      try {
+        return {
+          ...transfer,
+          assetInfo:
+            transfer.assetCategory === CryptoAssetCategories.fungible
+              ? await sip10AssetService.getAssetInfo(transfer.assetId, signal)
+              : await sip9AssetService.getAssetInfo(transfer.assetId, transfer.tokenValue!, signal),
+        };
+      } catch {
+        // if we're unable to find asset metadata, remove transfer
+        return undefined;
+      }
+    }
+  }
+
+  async function applyMarketData(activity: Activity, signal?: AbortSignal) {
+    if (
+      (activity.type === OnChainActivityTypes.sendAsset ||
```

Comment:
It would be more readable to add a few `const`s here. 

`const isFungible = activity.asset.category === CryptoAssetCategories.fungible`

```
if(isFungible)
```